#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/objects/floors/protos/Floor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/nvidia/jetbot/protos/JetBotRaspberryPiCamera.proto"

WorldInfo {
  basicTimeStep 16
  contactProperties [
    ContactProperties {
      material2 "steering_wheel"
      frictionRotation 1 1
    }
  ]
}
Viewpoint {
  orientation -0.49308100132475946 0.49928398748826736 0.7124511393565093 1.8916912886598203
  position -0.39596318294120736 -4.126383085866985 11.972233179691486
  follow "robot"
  followType "None"
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Floor {
  size 5 5
  tileSize 5 5
  appearance PBRAppearance {
    baseColorMap ImageTexture {
      url [
        "../textures/hard_track.png"
      ]
    }
    roughness 1
    metalness 0
  }
}
DEF ROBOT Robot {
  translation -1.94038 -0.280031 0.04
  rotation 0 0 -1 -1.5707953071795862
  children [
    JetBotRaspberryPiCamera {
      translation 0.11 0 0.07
      rotation 0 1 0 0.261799
    }
    DistanceSensor {
      translation 0.1 0.03 0
      rotation 0 0 1 0.3
      children [
        Shape {
          geometry DEF DS Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_left"
      boundingObject USE DS
      physics Physics {
      }
    }
    DistanceSensor {
      translation 0.1 -0.03 0
      rotation 0 0 1 -0.3
      children [
        Shape {
          geometry DEF DS Box {
            size 0.01 0.01 0.01
          }
        }
      ]
      name "ds_right"
      boundingObject USE DS
      physics Physics {
      }
    }
    DEF BODY Shape {
      appearance PBRAppearance {
        baseColor 0.7 0 0
        roughness 1
        metalness 0
      }
      geometry Box {
        size 0.2 0.08 0.05
      }
    }
    Hinge2Joint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor 0.075 0.0725 0
        minStop -0.5
        maxStop 0.5
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      device [
        RotationalMotor {
          name "left_steer"
        }
      ]
      device2 [
        RotationalMotor {
          maxTorque 0
        }
      ]
      endPoint DEF Wheel1 Solid {
        translation 0.075 0.0725 0
        rotation 1 1.3194274296012771e-14 6.636082735223269e-15 1.5707996938993234
        children [
          DEF WHEEL Shape {
            appearance PBRAppearance {
              baseColor 0 0.8 0
              baseColorMap ImageTexture {
                url [
                  "../textures/red_brick_wall.jpg"
                ]
              }
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.025
              radius 0.034
            }
          }
        ]
        name "solid(4)"
        contactMaterial "steering_wheel"
        boundingObject USE WHEEL
        physics DEF WHEEL_PH Physics {
        }
      }
    }
    Hinge2Joint {
      jointParameters HingeJointParameters {
        axis 0 0 1
        anchor 0.075 -0.0725 0
        minStop -0.5
        maxStop 0.5
      }
      jointParameters2 JointParameters {
        axis 0 1 0
      }
      device [
        RotationalMotor {
          name "right_steer"
        }
      ]
      device2 [
        RotationalMotor {
          maxTorque 0
        }
      ]
      endPoint DEF Wheel2 Solid {
        translation 0.075 -0.0725 0
        rotation 1 8.610348014768654e-15 1.5131740595510523e-14 1.57079969389938
        children [
          USE WHEEL
        ]
        name "solid(5)"
        contactMaterial "steering_wheel"
        boundingObject USE WHEEL
        physics DEF WHEEL_PH Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.075 0.0825 0
      }
      device [
        RotationalMotor {
          name "wheel3"
        }
      ]
      endPoint DEF Wheel3 Solid {
        translation -0.07500000000000001 0.08250000000000068 -5.510028035265896e-18
        rotation 1 -3.7730192466487225e-10 3.7729854897176924e-10 1.5707996938993893
        children [
          USE WHEEL
        ]
        name "solid(2)"
        boundingObject USE WHEEL
        physics DEF WHEEL_PH Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor -0.075 -0.0825 0
      }
      device [
        RotationalMotor {
          name "wheel4"
        }
      ]
      endPoint DEF Wheel4 Solid {
        translation -0.07499999999999996 -0.0825000000000029 -3.378333624605886e-17
        rotation 1 -7.617471317606674e-10 -9.073374784506339e-10 1.570799689155092
        children [
          USE WHEEL
        ]
        name "solid(3)"
        boundingObject USE WHEEL
        physics DEF WHEEL_PH Physics {
        }
      }
    }
  ]
  boundingObject USE BODY
  physics Physics {
  }
  controller "baseline"
  supervisor TRUE
}
